homeassistant/components/drop_connect/switch.py

```diff
@@ -0,0 +1,131 @@
+        icon="mdi:valve",
+        value_fn=lambda device: device.drop_api.water(),
+        set_fn=lambda device, value: device.set_water(value),
+    ),
+    DROPSwitchEntityDescription(
+        key=BYPASS_SWITCH,
+        translation_key=BYPASS_SWITCH,
+        icon="mdi:valve",
+        value_fn=lambda device: device.drop_api.bypass(),
+        set_fn=lambda device, value: device.set_bypass(value),
+    ),
+]
+
+# Defines which switches are used by each device type
+DEVICE_SWITCHES: dict[str, list[str]] = {
+    DEV_FILTER: [BYPASS_SWITCH],
+    DEV_HUB: [WATER_SWITCH, BYPASS_SWITCH],
+    DEV_PROTECTION_VALVE: [WATER_SWITCH],
+    DEV_SOFTENER: [BYPASS_SWITCH],
+}
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up the DROP switches from config entry."""
+    _LOGGER.debug(
+        "Set up switch for device type %s with entry_id is %s",
+        config_entry.data[CONF_DEVICE_TYPE],
+        config_entry.entry_id,
+    )
+
+    if config_entry.data[CONF_DEVICE_TYPE] in DEVICE_SWITCHES:
+        async_add_entities(
+            DROPSwitch(hass.data[DOMAIN][config_entry.entry_id], switch)
+            for switch in SWITCHES
+            if switch.key in DEVICE_SWITCHES[config_entry.data[CONF_DEVICE_TYPE]]
+        )
+
+
+class DROPSwitch(DROPEntity, SwitchEntity):
+    """Representation of a DROP switch."""
+
+    entity_description: DROPSwitchEntityDescription
+
+    def __init__(
+        self,
+        coordinator: DROPDeviceDataUpdateCoordinator,
+        entity_description: DROPSwitchEntityDescription,
+    ) -> None:
+        """Initialize the switch."""
+        super().__init__(entity_description.key, coordinator)
+        self.entity_description = entity_description
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return the state of the binary sensor."""
+        if self.entity_description.value_fn(self.coordinator) == 1:
+            return True
+        if self.entity_description.value_fn(self.coordinator) == 0:
+            return False
+        return None
+
+    async def async_turn_on(self, **kwargs: Any) -> None:
+        """Turn switch on."""
+        await self.entity_description.set_fn(self.coordinator, 1)
+        self._attr_is_on = True
+        self.async_write_ha_state()
+
+    async def async_turn_off(self, **kwargs: Any) -> None:
+        """Turn switch off."""
+        await self.entity_description.set_fn(self.coordinator, 0)
+        self._attr_is_on = False
+        self.async_write_ha_state()
+
+    @property
+    def icon(self) -> str:
+        """Return the icon to use for dynamic states."""
+        _icon = self.entity_description.icon or ""
+        if self.entity_description.key in (WATER_SWITCH, BYPASS_SWITCH):
+            if self._attr_is_on:
+                _icon = "mdi:valve-open"
+            else:
+                _icon = "mdi:valve-closed"
+        return _icon
```

Comment:
```suggestion
    @property
    def icon(self) -> str:
        """Return the icon to use for dynamic states."""
        if self.entity_description.key in {WATER_SWITCH, BYPASS_SWITCH}:
            if self.is_on:
                return "mdi:valve-open"
            return "mdi:valve-closed"
        return self.entity_description.icon
```

IMO we should try to set this during the coordinator update.

Reply:
To add, there are only 2 description types, why do we even fallback to entity description icon

Reply:
I used this approach specifically so I could get 100% coverage in my unit test.  At the moment, all of the switches have dynamic icons, but I didn't want to make the assumption that would always be the case.  By setting a return value and then optionally overriding it I don't miss the final return value with the two switches I'm currently implementing.  I know it's slightly odd, but can I leave this one alone so I can keep my 100% coverage score?

Reply:
> To add, there are only 2 description types, why do we even fallback to entity description icon

We have plans in the works for adding additional device types that will have switches without dynamic icons.  The fallback is just planning ahead for future additions.

Reply:
I went ahead and simplified the icon property in 8ad8014c75e5574b29cf506bd7a03f402b738ccc just so the unnecessary fallback case isn't an issue here.